### DISCUSSION TITLE
Fundamentos Entity Framework - Duvidas com Include

### DISCUSSION THREAD

**Discussion by NadiaTeles11** [balta-io/forum#351] [Tira Dúvidas] [answered]
Boa tarde.

Eu fiz o curso **Fundamentos Entity Framework** e surgiu uma dúvida.

Veja um exemplo abaixo:

**Repository**
 
```

public async Task<IQueryable<Area>> ObterTodos()
        {           
                var query =  _context.Area                   
                    .Include(x => x.Bloqueios)
                    .Include(x => x.Elevadores)
                    .Include(x => x.ControleVisitas)
                    .Include(x => x.Niveis)
                    .Include(x => x.Setor)
                     .ThenInclude(x => x.Site)
                    .Include(x => x.AreasInterligadas)
                    .Include(x => x.AreaMae)
                    .DefaultIfEmpty()
                    .Where(AreaQueries.ObterTodos());

                return query;
            }                 
        }

```

**Chamada usando GrapgQL**

```
public async Task<IQueryable<NcArea>> ReparticaoArea([Service] IAreaRepository areaRepository)
        {
            return await areaRepository.ObterTodos();
        }

```
Em uma determinada consulta quero obter apenas o nome da área e nesse caso não precisaria de includes. Em outra consulta preciso apresentar todas as informações e os includes seriam necessários. Em outra consulta quero obter a Area e os bloqueios associados. 
É possível em tempo de execução, obter os parâmetros que chegam no IQueryable e assim definir dinamicamente os includes que quero chamar?
É possível em tempo de execução, obter os parâmetros que chegam no IQueryable e criar dinamicamente os filtros e ordenação e aplica-los ao banco de dados?

Alguma dica, material ou curso sobre isto?
 
 Obrigada
 
 Nádia

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia, @NadiaTeles11, como vai?

É sim, ficaria algo parecido com isso:

```csharp
public async Task<IQueryable<Area>> ObterTodos(params Expression<Func<Area, object>>[] includes)
{
   var query =  _context.Area;

    foreach (var item in includes)
        query = query.Include(item);

   query = query.DefaultIfEmpty()
                    query = query.Where(AreaQueries.ObterTodos());

                return query;
}

```

PS:. Dá pra fazer isso genérico e criar um método de extensão também.

**Reply by NadiaTeles11** [reply]
Agradeço muito seu feedback e dicas. Vou investir nesse assunto. 
Muito Obrigada!
